python/lsst/obs/metadata/metadata.py

```diff
@@ -0,0 +1,179 @@
+        translation : `str` or `numbers.Number`
+            Translated header value derived directly from the {fitsKey}
+            header value.
+        """
+        return translator
+
+    def __init__(self, name, bases, dct):
+        super().__init__(name, bases, dct)
+
+        if hasattr(self, "name") and self.name is not None:
+            MetadataTranslator.translators[self.name] = self
+
+        for standardKey, fitsKey in self._unitMap.items():
+            translator = self._makeUnitMapping(standardKey, fitsKey)
+            setattr(self, f"to{standardKey.capitalize()}", staticmethod(translator))
+
+        for standardKey, constant in self._constMap.items():
+            translator = self._makeConstMapping(standardKey, constant)
+            setattr(self, f"to{standardKey.capitalize()}", staticmethod(translator))
+
+
+class MetadataTranslator(metaclass=MetadataMeta):
+    """Per-instrument metadata translation support"""
+
+    _unitMap = {}
+    """Dict of one-to-one mappings for header translation from standard
+    property to corresponding FITS keyword."""
+
+    _constMap = {}
+    """Dict defining a constant for specified standard properties."""
+
+    translators = dict()
+    """All registered metadata translation classes."""
+
+    supportedInstrument = None
+    """Name of instrument understood by this translation class."""
+
+    @classmethod
+    @abstractmethod
+    def canTranslate(cls, header):
+        """Indicate whether this translation class can translate the
+        supplied header.
+
+        Parameters
+        ----------
+        header : `dict`-like
+           Header to convert to standardized form.
+
+        Returns
+        -------
+        can : `bool`
+            `True` if the header is recognized by this class. `False`
+            otherwise.
+        """
+        raise NotImplementedError()
+
+
+class VisitInfo:
+    """Standardized representation of an instrument FITS header.
+
+    Parameters
+    ----------
+    header : `dict`-like
+        Representation of an instrument FITS header accessible as a `dict`.
+    translator : `MetadataTranslator`-class, `optional`
+        If not `None`, the class to use to translate the supplied headers
+        into standard form. Otherwise each registered translator class will
+        be asked in turn if it knows how to translate the supplied header.
+
+    Raises
+    ------
+    ValueError
+        The supplied header was not recognized by any of the registered
+        translators.
+    """
+
+    def __init__(self, header, translator=None):
+        if translator is None:
+            for name, trans in MetadataTranslator.translators.items():
+                if trans.canTranslate(header):
+                    log.debug(f"Using translation class {name}")
+                    translator = trans
+                    break
+            else:
+                raise ValueError("None of the registered translation classes understood this header")
```

Comment:
Finding a matching translator for a header seems like something best factored out into a separate function rather than implemented directly in `__init__`.